pkg/controller/deployment.go

```
@@ -604,3 +611,65 @@ func (dc *controller) updateMachineDeploymentFinalizers(ctx context.Context, mac
 		dc.updateMachineDeploymentFinalizers(ctx, machineDeployment, finalizers)
 	}
 }
+
+func (dc *controller) setMachinePriorityAnnotationAndUpdateTriggeredForDeletion(ctx context.Context, mcd *v1alpha1.MachineDeployment) error {
+	var toBeDeletedMachineNames, toBeSkippedMachineNames []string
+	toBeDeletedMachineNames = annotations.GetMachineNamesTriggeredForDeletion(mcd)
+	if len(toBeDeletedMachineNames) == 0 {
+		return nil
+	}
+	klog.V(3).Infof("MachineDeployment %q has #%d machine(s) marked for deletion, toBeDeletedMachineNames=%v", mcd.Name, len(toBeDeletedMachineNames), toBeDeletedMachineNames)
+	for _, machineName := range toBeDeletedMachineNames {
+		mc, err := dc.machineLister.Machines(dc.namespace).Get(machineName)
+		if apierrors.IsNotFound(err) {
+			klog.V(4).Infof("Machine %q is not found in MachineDeployment %q - skip setting MachinePriority=1 annotation", machineName, mcd.Name)
+			toBeSkippedMachineNames = append(toBeSkippedMachineNames, machineName)
+			continue
+		}
+		if machineutils.IsMachineFailedOrTerminating(mc) {
+			klog.V(4).Infof("Machine %q of MachineDeployment %q is in Failed/Terminating state - skip setting MachinePriority=1 annotation", machineName, mcd.Name)
+			toBeSkippedMachineNames = append(toBeSkippedMachineNames, machineName)
+			continue
+		}
+		if mc.Annotations[machineutils.MachinePriority] == "1" {
+			klog.V(4).Infof("Machine %q of MachineDeployment %q already marked with MachinePriority=1 annotation", machineName, mcd.Name)
+			continue
+		}
+		mcAdjust := mc.DeepCopy()
+		mcAdjust.Annotations[machineutils.MachinePriority] = "1"
+		_, err = dc.controlMachineClient.Machines(mcAdjust.Namespace).Update(ctx, mcAdjust, metav1.UpdateOptions{})
+		if err != nil {
+			klog.Errorf("Failed to set MachinePriority=1 annotation on Machine %q of MachineDeployment %q: %v", machineName, mcd.Name, err)
+			//TODO: Return error here or just log and continue for best-effort?
```

i think returning an error would be fine. The call to update the Machine can fail for 2 possible reasons:

1. Client side throttling to the Kube Apiserver
2. Conflict (attempt to update a stale object)

In case of (1) if error is not returned then it will also fail for other machines (in the loop) as well which is unnecessary calls to KAPI. But in case of (2) it might work for other machines. So either we make it complicated and differentiate based on error code and return or not return an error based on that. Alternatively we keep it simple and always return on first error.

---

Ok, my main concern was that the first encountered error for 2) causes all the subsequent Machines to fail priority-1 annotation.  But I didn't consider that client-side throttling would anyways fail the subsequent guys. 

Either I make it more complex and distinguish throttling errors from update errors or just keep it simple as is.  The reconcile will be attempted again anyways if an error is returned.

---

Ig we agree that a simple way out like we have now is apt, so we can remove this TODO